dask/base.py

```diff
@@ -377,36 +379,69 @@ def optimization_function(x):
     return getattr(x, "__dask_optimize__", dont_optimize)
 
 
-def collections_to_dsk(collections, optimize_graph=True, optimizations=(), **kwargs):
+def newstyle_collections(collections):
+    from dask.delayed import Delayed
+
+    is_newstyle = [
+        isinstance(c, DaskCollection2) and not isinstance(c, Delayed)
+        for c in collections
+    ]
+    if any(is_newstyle):
+        if not all(is_newstyle):
+            raise RuntimeError("Provided new- and old-style collections.")
+        if not all(
+            isinstance(c.__dask_graph_factory__(), TaskGraphFactory)
+            for c in collections
+        ):
+            raise TypeError("Newstyle collections must have a TaskGraphFactory graph.")
+
+        return True
+    return False
+
+
+def collections_to_dsk(
+    collections, optimize_graph=True, optimizations=(), **kwargs
+) -> TaskGraphFactory:
```

Comment:
The code here is still pretty ugly since it requires boilerplate all over for compat purposes.

It probably doesn't make sense to have a switch inside of this function. I'd likely define a function that only works for newstyle and oldstyle collections since we'll need another switch outside as well.